New script to automate job suspension/unsuspension: suspendmanager

plule opened the pull request:
Add `suspendmanager`, addressing most of the issue https://github.com/DFHack/dfhack/issues/2955

It is intended to work as a replacement for `autounsuspend` with the additional functionality of also suspending jobs in way that it prevents construction jobs to block each other. It works by computing a "risk of being stuck" factor for a job, based on how close it is to be enclosed. Jobs that are neighbouring another job more at risk of being stuck will be suspended.

Modified:
- `unsuspend` now support `--skipblocking`, to not unsuspend jobs that risk blocking other

Added:
- `suspend [--onlyblocking]`: Suspending all jobs, or only the one that risk blocking other
- `suspendmanager`: Automation of `suspend`/`unsuspend` intended to remove job suspension micromanagement, reusable module for other suspension management tools

Misc:
- `unsuspend --quiet` is now quieter, previously it was only removing the "0 job" messages, now it removes all messages
- `unsuspend` now uses argparse

Configuration GUI:
![image](https://user-images.githubusercontent.com/630159/222914484-881de451-3510-463f-a22f-94fdbe5a2e37.png)

This PR only lacks the migration from `autounsuspend`, which is currently left untouched. I don't really now how to do this

- [x] `suspend.lua`
- [x] `suspendmanager.lua` -- dumb trigger
- [x] `gui/suspendmanager.lua`
- [x] rework `suspendmanager` triggers
- [x] `unsuspend` skips blocking jobs
- [x] doc/terminology cleanup

myk002 commented on smartsuspend.lua:
```diff
@@ -0,0 +1,117 @@
+    if jobShouldBeSuspended(job) then
+        suspend(job)
+    else
+        job.flags.suspend = false
```
this script shouldn't be unsuspending jobs unless it's taking over the job of autounsuspend (which I recommend that it does -- I think this functionality should be part of `autounsuspend`, on by default, and controllable with a setting)

plule replied:
Yep, it makes sense

plule commented on smartsuspend.lua:
```diff
@@ -0,0 +1,117 @@
+--- Check if a building will block the path
+--- @param building building|building_constructionst|nil
+--- @return boolean
+local function isBlockingConstruction(building)
```
Ha, dark knowledge of dfhack, thanks

plule commented:
@myk002 regarding the best way to call this from: Calling it from the main "unsuspend" loop, where it also suspends jobs works well, but at the condition that the `autounsuspend` frequency is lowered from 1 day to 10 ticks. I don't realize how acceptable it is, since that's a pretty big jump.

Alternatively, I can investigate the event route, to try to hook to the "job" completed event, though that's uncharted territory for me

plule commented:
Welll the gui work went smoother than I expected, so I directly added the integration in unsuspend and autounsuspend.

For now it's going the route of running it every 10 ticks though, which is maybe a problem

myk002 commented on autounsuspend.lua:
```diff
@@ -8,19 +8,35 @@ local persist = require('persist-table')
 local GLOBAL_KEY = 'autounsuspend' -- used for state change hooks and persistence
 
 enabled = enabled or false
+smart = smart or nil
+if smart == nil then
+    smart = true
+end
```
```suggestion
smart = smart == nil and true or smart
```

myk002 commented on docs/autounsuspend.rst:
```diff
@@ -16,3 +16,12 @@ Usage
+-------
+
+``enable_smart``
+    Automatically suspend jobs that risk blocking other construction jobs. See `smartsuspend`.
```
should mention that this is the default

myk002 commented on gui/autounsuspend.lua:
```diff
@@ -0,0 +1,77 @@
+end
+
+function Autounsuspend:onRenderBody()
+    if self.next_refresh_ms <= dfhack.getTickCount() then
+        self:refresh_data()
+    end
```
unless you're reporting on the live number of suspended jobs, there's really no reason to have an autorefresh.

plule replied:
oops forgot to adress this one

plule replied:
And done this time

myk002 commented on unsuspend.lua:
```diff
@@ -184,9 +185,20 @@ if dfhack_flags.module then
 end
 
-local unsuspended_count, flow_count, buildingplan_count = 0, 0, 0
+local opts = utils.invert{...}
+local quiet = opts['-q'] or opts['--quiet']
+local smart = opts['-s'] or opts['--smart']
```
this is *ok* for now, but at some point you should just use a real commandline parsing function like `argparse.processArgsGetopt()` and avoid having to do the matching and parsing yourself. docs here: https://docs.dfhack.org/en/latest/docs/dev/Lua%20API.html#argparse

myk002 commented on docs/unsuspend.rst:
```diff
@@ -19,6 +22,15 @@ Usage
+-------
+
+``-q``, ``--quiet``
+    Don't report the number of unsuspended tiles
```
jobs, not tiles, right?

myk002 commented on autounsuspend.lua:
```diff
@@ -8,19 +8,35 @@ local persist = require('persist-table')
+            table.insert(args, '--smart')
+        end
+        dfhack.run_script('unsuspend', table.unpack(args))
+        dfhack.timeout(10, 'ticks', event_loop)
```
yeah, this is *really* frequent, even more frequent than `autonestbox`, which has similar requirements for "mark it before someone takes it". I'd be concerned about running even cpp code at this frequency. Registering a job completion handler might be the better approach.

plule replied:
I switched to triggering on job creation and completion, and it works as expected :)

myk002 reviewed (commented):
I think I need to discuss with the rest of the team about whether these are the appropriate dependency chains. does `smartsuspend` *need* to be a new, separate script, or can it be a mode of `unsuspend`? or should it just be called `suspend` and maybe take on more features about selecting jobs for suspension? Maybe we just need to rename `autounsuspend` to `suspendmanager` or something to make it clearer that it can now both suspend *and* unsuspend, as appropriate.

myk002 commented:
ok, I wrote a plan at DFHack/dfhack#2955. Thoughts?

plule commented:
It makes sense to me, just like your other suggestions. I should be able to look at this again on Monday.

plule commented on gui/control-panel.lua:
```diff
@@ -30,6 +30,7 @@ local FORT_SERVICES = {
     'prioritize',
     'seedwatch',
     'starvingdead',
+    'suspendmanager',
```
Sure, done (still a bit uneasy about disabling the previous stuff)

plule commented on docs/suspendmanager.rst:
```diff
@@ -0,0 +1,26 @@
+
+::
+
+    enable suspendmanager
```
`suspendmanager status` is now removed, it now under `suspendmanager`

plule commented on suspendmanager.lua:
```diff
@@ -0,0 +1,155 @@
+local function run_now()
+    if prevent_blocking then
+        dfhack.run_script('suspend', '--onlyblocking')
+        dfhack.run_script('unsuspend', '--quiet', '--skipblocking')
```
Done, with a single loop. For the incremental mode the fact that `job.pos` behave weirdly gets in the way, maybe I'll try again to do it in a further PR, but I'm a bit afraid to introduce bugs right now.

plule commented:
@myk002 The latest push addresses all the remaining comments. I did a change from the initial plan, which is to move all the reusable bits in `suspendmanager` instead of having them scattered between `suspend` and `unsuspend`, as it was getting weird (`unsuspend` needed `suspend` to check for blocking jobs, and `suspend` could arguably use `unsuspend` for the underwater and buildingplan checks). It makes a nice pattern where `suspendmanager` is the main entrypoint with all these smart checks there, while `suspend` and `unsuspend` are really basic glue code for some of its features.

I was also able to confirm that it behaves correctly with `buildingplan` with the latest (and awesomest) DFHack release.

plule commented on docs/unsuspend.rst:
```diff
@@ -19,6 +19,15 @@ Usage
 
     unsuspend
 
+Options
```
You have sharp eyes! Done

myk002 commented on changelog.txt:
```diff
@@ -15,6 +15,8 @@ that repo.
 ## New Scripts
 - `suspendmanager`: Automatic job suspension management (replaces `autounsuspend`)
+- `gui/suspendmanager`: Graphical configuration interface for `suspendmanager`
+- `suspend`: uspends building construction jobs
```
```suggestion
- `suspend`: suspends building construction jobs
```
